src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperAttributeList.cs

```diff
@@ -0,0 +1,231 @@
+        /// </param>
+        /// <returns>The first <see cref="TagHelperAttribute"/> with <see cref="TagHelperAttribute.Name"/> matching
+        /// <paramref name="name"/>.
+        /// </returns>
+        /// <remarks><paramref name="name"/> is compared case-insensitively. When setting,
+        /// <see cref="TagHelperAttribute"/>s <see cref="TagHelperAttribute.Name"/> must be <c>null</c> or
+        /// case-insensitively match the specified <paramref name="name"/>.</remarks>
+        /// <example>
+        /// <code>
+        /// var attributes = new TagHelperAttributeList();
+        ///
+        /// // Will "value" be converted to a TagHelperAttribute with a null Name
+        /// attributes["name"] = "value";
+        ///
+        /// // TagHelperAttribute.Name must match the specified name.
+        /// attributes["name"] = new TagHelperAttribute("name", "value");
+        /// </code>
+        /// </example>
+        public new TagHelperAttribute this[[NotNull] string name]
+        {
+            get
+            {
+                return base[name];
+            }
+            [param: NotNull]
+            set
+            {
+                // Name will be null if user attempts to set the attribute via an implicit conversion:
+                // output.Attributes["someName"] = "someValue"
+                if (value.Name == null)
+                {
+                    value.Name = name;
+                }
+
+                var attributeReplaced = false;
+
+                for (var i = 0; i < Attributes.Count; i++)
+                {
+                    if (NameEquals(name, Attributes[i]))
+                    {
+                        // We replace the first attribute with the provided value, remove all the rest.
+                        if (!attributeReplaced)
+                        {
+                            // We replace the first attribute we find with the same name.
+                            Attributes[i] = value;
+                            attributeReplaced = true;
+                        }
+                        else
+                        {
+                            Attributes.RemoveAt(i--);
+                        }
+                    }
+                }
+
+                // If we didn't replace an attribute value we should add value to the end of the collection.
+                if (!attributeReplaced)
+                {
+                    if (!NameEquals(name, value))
+                    {
+                        throw new ArgumentException(
+                            Resources.FormatTagHelperAttributeList_CannotAddAttribute(
+                                nameof(TagHelperAttribute),
+                                nameof(TagHelperAttribute.Name),
+                                value.Name,
+                                name),
+                            nameof(name));
+                    }
```

Comment:
more generally, `// Replaces A with B.` does not describe the `attributes["A"]` line correctly because the setter removes all "A" entries but no existing "B" entries.